daemon/containerd/image_snapshot.go

```diff
@@ -63,22 +61,151 @@ func (i *ImageService) PrepareSnapshot(ctx context.Context, id string, parentIma
 	ls := i.client.LeasesService()
 	lease, err := ls.Create(ctx, leases.WithID(id))
 	if err != nil {
-		return err
+		return nil, err
 	}
 	ctx = leases.WithLease(ctx, lease.ID)
 
-	snapshotter := i.client.SnapshotService(i.StorageDriver())
+	if err := i.prepareInitLayer(ctx, id, parentSnapshot, initFunc); err != nil {
+		return nil, err
+	}
 
-	if err := i.prepareInitLayer(ctx, id, parentSnapshot, setupInit); err != nil {
-		return err
+	sn := i.client.SnapshotService(i.StorageDriver())
+	layer := &rwLayer{
+		id:              id,
+		snapshotterName: i.StorageDriver(),
+		snapshotter:     sn,
+		refCountMounter: i.refCountMounter,
+		lease:           lease,
 	}
 
 	if !i.idMapping.Empty() {
-		return i.remapSnapshot(ctx, snapshotter, id, id+"-init")
+		return layer, i.remapSnapshot(ctx, sn, id, id+"-init")
+	}
+
+	_, err = sn.Prepare(ctx, id, id+"-init")
+	return layer, err
+}
+
+type rwLayer struct {
+	id              string
+	snapshotter     snapshots.Snapshotter
+	snapshotterName string
+	refCountMounter snapshotter.Mounter
+	root            string
+	lease           leases.Lease
+}
+
+func (l *rwLayer) mounts(ctx context.Context) ([]mount.Mount, error) {
+	return l.snapshotter.Mounts(ctx, l.id)
+}
+
+func (l *rwLayer) Mount(mountLabel string) (string, error) {
+	ctx := context.TODO()
+
+	// TODO: Investigate how we can handle mountLabel
+	_ = mountLabel
+	mounts, err := l.mounts(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	var root string
+	if root, err = l.refCountMounter.Mount(mounts, l.id); err != nil {
+		return "", fmt.Errorf("failed to mount %s: %w", root, err)
+	}
+	l.root = root
+
+	log.G(ctx).WithFields(log.Fields{"container": l.id, "root": root, "snapshotter": l.snapshotterName}).Debug("container mounted via snapshotter")
+	return root, nil
+}
+
+// GetLayerByID returns a layer by ID
+// called from daemon.go Daemon.restore().
+func (i *ImageService) GetLayerByID(cid string) (container.RWLayer, error) {
+	ctx := context.TODO()
+
+	sn := i.client.SnapshotService(i.StorageDriver())
+	if _, err := sn.Stat(ctx, cid); err != nil {
+		if !cerrdefs.IsNotFound(err) {
+			return nil, fmt.Errorf("failed to stat snapshot %s: %w", cid, err)
+		}
+		return nil, errdefs.NotFound(fmt.Errorf("RW layer for container %s not found", cid))
 	}
 
-	_, err = snapshotter.Prepare(ctx, id, id+"-init")
-	return err
+	ls := i.client.LeasesService()
+	lss, err := ls.List(ctx, "id=="+cid)
+	if err != nil {
+		return nil, err
+	}
+
+	var lease leases.Lease
+	switch len(lss) {
+	case 1:
+		lease = lss[0]
+	case 0:
+		log.G(ctx).WithField("container", cid).Warn("rw layer lease not found for container")
```

Comment:
Is it safe to proceed without a lease, or should we return an error?

Reply:
Hmmm, technically the snapshot exists, so it can be used, but without the lease there's no guarantee that it won't get GC'd.
I think it makes sense to error out in this case.